Tools/autotest/arduplane.py

```diff
@@ -1834,6 +1834,167 @@ def fly_terrain_mission(self):
+        self.set_parameter("SIM_GPS2_DISABLE", 0)
+        self.set_parameter("SIM_BARO2_DISABL", 0)
+        self.set_parameter("SIM_BARO_COUNT", 2)
+        self.set_parameter("ARSPD2_TYPE", 2)
+        self.set_parameter("ARSPD2_USE", 1)
+        self.set_parameter("ARSPD2_PIN", 2)
+        self.set_parameter("SIM_SPEEDUP", 20)
+        self.set_parameter("SIM_RATE_HZ", 200)
+
+        # some parameters need reboot to take effect
+        self.reboot_sitl()
+
+        self.lane_switches = []
+        # add an EKF lane switch hook
+        def statustext_hook(mav, message):
+            if message.get_type() != 'STATUSTEXT':
+                return
+            # example msg: EKF3 lane switch 1
+            if not message.text.startswith("EKF3 lane switch "):
+                return
+            newlane = int(message.text[-1])
+            self.lane_switches.append(newlane)   
+        self.install_message_hook(statustext_hook)
+
+        # get flying
+        self.takeoff(alt=75)
+        self.change_mode('CIRCLE')
+
+        try:
+            ###################################################################################################
+            self.progress("Checking EKF3 Lane Switching trigger from all sensors")
+            ###################################################################################################
+            self.start_subtest("ACCELEROMETER: Change z-axis offset")
+            # create an accelerometer error by changing the Z-axis offset
+            old_parameter = self.get_parameter("INS_ACCOFFS_Z")
+            self.set_parameter("INS_ACCOFFS_Z", old_parameter + 5)
+            self.wait_statustext(text="EKF3 lane switch", timeout=30)
+            if self.lane_switches != [1]:
+                raise NotAchievedException("Expected lane switch 1, got %s" % str(self.lane_switches))
+            # Cleanup
+            self.set_parameter("INS_ACCOFFS_Z", old_parameter)
+            self.lane_switches = []
+            self.wait_heading(0, accuracy=10, timeout=60)
+            self.wait_heading(180, accuracy=10, timeout=60)
+            ###################################################################################################
+            self.start_subtest("BAROMETER: Freeze to last measured value")
+            old_parameter = self.get_parameter("SIM_BARO2_FREEZE")
+            # create a barometer error by inhibiting any pressure change while changing altitude
+            self.set_parameter("SIM_BARO2_FREEZE", 1)
+            self.set_rc(2, 2000)
+            self.delay_sim_time(2)
+            self.wait_statustext(text="EKF3 lane switch", timeout=30)
+            if self.lane_switches != [0]:
+                raise NotAchievedException("Expected lane switch 0, got %s" % str(self.lane_switches))
+            # Cleanup
+            self.set_rc(2, 1500)
+            self.set_parameter("SIM_BARO2_FREEZE", old_parameter)
+            self.lane_switches = []
```

Comment:
I know I suggested this, but I realise now this introduces a race condition by clearing the list of switches. If we got two lane switches in one test then we would pass when we should fail.
We should remove the "self.lane_switches = []" after each test, and instead check for the accumulated set of lane switches, so the 2nd test would be:
 if self.lane_switches = [1,0]
then next one would be
 if self.lane_switches = [1,0,1]
etc

Reply:
I see, I'll make that change and test it.